File: breadth-first-search/LC107.java
import java.util.*;

public class LC107
{
	public static void main(String[] args)
	{
	}

	// 107. Binary Tree Level Order Traversal II
	static List<List<Integer>> levelOrderBottom(TreeNode root)
	{
		List<List<Integer>> list = new ArrayList<List<Integer>>();
		if (root == null)
			return list;
		Deque<TreeNode> queue = new ArrayDeque<TreeNode>();
		queue.addLast(root);
		int count = 1;
		while (!queue.isEmpty())
		{
			int hold = 0;
			List<Integer> level = new ArrayList<Integer>();
			for (int i = 0; i < count; i++)
			{
				TreeNode node = queue.removeFirst();
				level.add(node.val);
				if (node.left != null)
				{
					queue.addLast(node.left);
					hold++;
				}
				if (node.right != null)
				{
					queue.addLast(node.right);
					hold++;
				}
			}
			list.add(level);
			count = hold;
		}
		for (int i = 0; i < list.size() - 1 - i; i++)
		{
			int len = list.size();
			List<Integer> save = list.get(i);
			list.set(i, list.get(len - 1 - i));
			list.set(len - 1 - i, save);
		}
		return list;    
	}

	static class TreeNode
	{
		int val;
		TreeNode left;
		TreeNode right;
		TreeNode(int x) { val = x; }
	}
}

File: python/tree/lc107.py
"""107. Binary Tree Level Order Traversal II

Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its bottom-up level order traversal as:
[
  [15,7],
  [9,20],
  [3]
]
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
  def levelOrderBottom(self, root):
    """
    :type root: TreeNode
    :rtype: List[List[int]]
    """
    if not root:
      return []
    
    # Exactly the same as LC 102 Binary Tree Level Order Traversal II
    # The only difference is to reverse the list of per-level sequences
    curr_level = [root]
    next_level = []
    levels = []
    
    while curr_level:
      vals = []
      for node in curr_level:
        vals.append(node.val)
        if node.left:
          next_level.append(node.left)
        if node.right:
          next_level.append(node.right)
          
      levels.append(vals)    
      curr_level = next_level    
      next_level = []
    
    # reverse the list of per-level sequences
    return reversed(levels)  
